File: CsPlugin/Content/Scripts/Cs/Coroutine/Types_Coroutine.ts
// Copyright 2017-2023 Closed Sum Games, LLC. All Rights Reserved.
// MIT License: https://opensource.org/license/mit/
// Free for use and distribution: https://github.com/closedsum/core

/// <reference path="../../typings/ue.d.ts">/>
// ; typing info for auto-completion in Visual Studio Code

// require('Cs/Coroutine/Types_Coroutine.js')

declare namespace NJsCoroutine {
    var EState:
    {
        Free:       { Value: 0, Name: "Free" },
        Init:       { Value: 1, Name: "Init" },
        Update:     { Value: 2, Name: "Update" },
        End:        { Value: 3, Name: "End" },
        EState_MAX: { Value: 4, Name: "EState_MAX" }
    };
    var EMessage:
    {
        Notify:       { Value: 0, Name: "Notify" },
        Listen:       { Value: 1, Name: "Listen" },
        Abort:        { Value: 2, Name: "Abort" },
        EMessage_MAX: { Value: 3, Name: "EMessage_MAX" }
    };
    var EEndReason:
    {
        EndOfExecution: { Value: 0, Name: "EndOfExecution" },
        AbortMessage:   { Value: 1, Name: "AbortMessage" },
        AbortCondition: { Value: 2, Name: "AbortCondition" },
        OwnerIsInvalid: { Value: 3, Name: "OwnerIsInvalid" },
        Parent:         { Value: 4, Name: "Parent" },
        UniqueInstance: { Value: 5, Name: "UniqueInstance" },
        Shutdown:       { Value: 6, Name: "Shutdown" },
        Manual:         { Value: 7, Name: "Manual" },
        EEndReason_MAX: { Value: 8, Name: "EEndReason_MAX" }
    };

    class FOwner {
        Owner: object;
        UEObject: UObject;
        bObject: boolean;
        UEActor: Actor;
        IsObject(): boolean;
        SetObject(o: object): void;
        Copy(from: NJsCoroutine.FOwner): void;
        Reset(): void;
    }
}

//declare type NJsCoroutine_EState = NJsCoroutine.EState;
//declare type NJsCoroutine_EMessage = NJsCoroutine.EMessage;
//declare type NJsCoroutine_EEndReason = NJsCoroutine.EEndReason;
declare type NJsCoroutine_FOwner = NJsCoroutine.FOwner;

declare namespace NJsCoroutine.NRegister {
    class FInfo {
        Index: number
    }

    class FMap {
        Infos: NJsCoroutine.NRegister.FInfo[];
        Values: any[];
        UseValues: boolean[];
        SetUsedValue(index: number): void;
        SetValue(i: number, v: any): void;
        GetValue(i: number): any;
        Reset(): void;
    }
}

declare type NJsCoroutine_NRegister_FInfo = NJsCoroutine.NRegister.FInfo;
declare type NJsCoroutine_NRegister_FMap = NJsCoroutine.NRegister.FMap;

declare namespace NJsCoroutine.NPayload {
    class FImpl {
        Index: number;;
        Group : ECsUpdateGroup;
        CoroutineImpl: IterableIterator<any>;
        StartTime: CsTime;
        Owner: NJsCoroutine.FOwner;
        AbortImpls: ((any) => boolean)[];
        OnAborts: ((any) => void)[];
        OnEnds: ((any) => void)[];
        ParentHandle: CsRoutineHandle;
        bDoInit: boolean;
        bPerformFirstUpdate: boolean;
        AbortMessages: string[];
        RegisterMap: NJsCoroutine.NRegister.FMap;
        Name: string;
        SetIndex(index: number): void;
        GetIndex(): number;
        SetValue(index: number, value: any): void;
        Reset(): void;
        Copy(from: NJsCoroutine.NPayload.FImpl): void;
        IsValidChecked(context: string): boolean;
    }
}

declare type NJsCoroutine_NPayload_FImpl = NJsCoroutine.NPayload.FImpl;

File: CsPlugin/Content/Python/Cs/Coroutine/Types_Coroutine.py
# Copyright 2017-2024 Closed Sum Games, LLC. All Rights Reserved.
# MIT License: https://opensource.org/license/mit/
# Free for use and distribution: https://github.com/closedsum/core

import unreal as ue
from enum import Enum
import copy

# Library
# - Cs/Library/Library_Common.py
import Cs.Library.Library_Common as Cs_Library_Common

INDEX_NONE = -1

ROUTINE_VALUE_SIZE = 16

# "typedefs" - library (py)
CommonLibrary = Cs_Library_Common.NPyCommon.FLibrary

# "typedefs" - functions
check           = CommonLibrary.check
checkf          = CommonLibrary.checkf
IsValidObject   = CommonLibrary.IsValidObject

class NPyCoroutine:
    class EState(Enum): 
        Free = 0
        Init = 1
        Update = 2
        End = 3
        EState_MAX = 4

    class EMessage(Enum):
        Notify = 0
        Listen = 1
        Abort = 2
        EMessage_MAX = 3

    class EEndReason(Enum):
        EndOfExecution = 0
        AbortMessage = 1
        AbortCondition = 2
        OwnerIsInvalid = 3
        Parent = 4
        UniqueInstance = 5
        Shutdown = 6
        Manual = 7
        EEndReason_MAX = 8

    class FOwner:
        class NCached:
            class NStr:
                SetObject: str = "NPyCoroutine.FOwner.SetObject"
        def __init__(self):
            self.Owner: object = None
            self.UEObject: ue.Object = None
            self.bObject: bool = False
            self.UEActor: ue.Actor = None

        def IsObject(self) -> bool:
             return self.bObject

        def SetObject(self, o: any):
            context: str = NPyCoroutine.FOwner.NCached.NStr.SetObject

            self.Owner = o

            if (isinstance(o, ue.Object)):
                self.UEObject = o
                self.bObject = True

            checkf(self.bObject, context + ": o is NOT a UObject.")

            if (isinstance(o, ue.Actor)):
                self.UEActor = o

        def Copy(self, b: any):
            checkf(isinstance(b, NPyCoroutine.FOwner), "b is NOT of type: NPyCoroutine.FOwner")
            
            self.Owner    = b.Owner
            self.UEObject = b.UEObject
            self.bObject  = b.bObject
            self.UEActor  = b.UEActor

        def Reset(self):
            self.Owner    = None
            self.UEObject = None
            self.bObject  = False
            self.UEActor  = None

    class NRegister:
        class FInfo:
            def __init__(self):
                self.Index = INDEX_NONE

        class FMap:
            def __init__(self):
                self.Infos = []

                self.Values     = [None] * ROUTINE_VALUE_SIZE
                self.UsedValues = [False] * ROUTINE_VALUE_SIZE

            def SetUsedValue(self, index: int):
                if (self.UsedValues[index] == False):
                    info        = NPyCoroutine.NRegister.FInfo()
                    info.Index  = index;
                   
                    self.Infos.append(info)

                    self.UsedValues[index] = True
                    
            def SetValue(self, i: int, v: any):
                self.SetUsedValue(i)
                self.Values[i] = v

            def GetValue(self, i: int) -> any:
                self.SetUsedValue(i)
                return self.Values[i]

            def Reset(self):
                for info in self.Infos:
                    self.UsedValues[info.Index] = False
                self.Infos = []

    class NPayload:
        class FImpl:
            def __init__(self):
                self.Index = INDEX_NONE

                UpdateGroupLibrary = ue.CsScriptLibrary_UpdateGroup
                self.Group         = UpdateGroupLibrary.get_max()

                self.CoroutineImpl = None
                self.StartTime = ue.CsTime()
                self.Owner = NPyCoroutine.FOwner()
                self.AbortImpls = []
                self.OnAborts = []
                self.OnEnds = []
                self.ParentHandle = ue.CsRoutineHandle()
                self.bDoInit = True
                self.bPerformFirstUpdate = True
                self.AbortMessages = []
                self.RegisterMap = NPyCoroutine.NRegister.FMap()
                self.Name = ""

            def SetIndex(self, index: int):
                 self.Index = index

            def GetIndex(self):
                return self.Index

            def SetValue(self, index: int, value: any):
                self.RegisterMap.SetValue(index, value)

            def Reset(self):
                self.CoroutineImpl = None
                self.StartTime = ue.CsTime()
                self.Owner.Reset()
                self.AbortImpls = []
                self.OnAborts = []
                self.OnEnds = []
                self.ParentHandle = ue.CsRoutineHandle()
                self.bDoInit = True
                self.bPerformFirstUpdate = True
                self.AbortMessages = []
                self.RegisterMap.Reset()
                self.Name = ""

            def Copy(self, b: any):
                checkf(isinstance(b, NPyCoroutine.NPayload.FImpl), "b is NOT of type: NPyCoroutine.NPayload.FImpl")

                self.Index = b.Index
                self.Group = b.Group
                self.CoroutineImpl = b.CoroutineImpl
                self.StartTime = copy.deepcopy(b.StartTime)
                self.Owner = b.Owner
                self.AbortImpls = copy.deepcopy(b.AbortImpls)
                self.OnAborts = copy.deepcopy(b.OnAborts)
                self.OnEnds = copy.deepcopy(b.OnEnds)
                self.ParentHandle = copy.deepcopy(b.ParentHandle)
                self.bDoInit = b.bDoInit
                self.bPerformFirstUpdate = b.bPerformFirstUpdate
                self.AbortMessages = copy.deepcopy(b.AbortMessages)
                # NOTE: Should create / use a copy function
                self.RegisterMap = copy.deepcopy(b.RegisterMap)
                self.Name = b.Name

            def IsValidChecked(self, context: str) -> bool:
                # TODO: Check self.CoroutineImpl is callable
                #check(isinstance(self.CoroutineImpl, types.GeneratorType))
                return True